docs/proposals/hosted_components_templates.md

```diff
@@ -0,0 +1,557 @@
+
+This allowed any changes for the installation to be maintained by the developers
+and keep them from needing to be fluent with Ansible. It also afforded a level
+of abstraction from needing to keep in sync with TSB related changes and then
+translate them to changes within an Ansible role.
+
+However, there is still the gap of ensuring our role configures changes appropriate
+for the version of the component we are installing. Prior to 3.4 with Aggregated
+Logging this was resolved and contained due to using a deployer pod, since everything
+was versioned and contained within an artifact the Integration Services team
+provided. This is something that is addressed again now with the Service Broker,
+APBs are configured in such a way that a containerized install/uninstall is done
+for the component.
+
+## Design
+We should describe a way to create a containerized playbook and role to mirror
+what is done with ABPs for hosted components so that the openshift-ansible roles
+would reuse these containers to install/uninstall these components.
+
+The containerized role would follow the pattern used by the TSB within its
+openshift-ansible role `template_service_broker` to maintain a lower learning
+curve for those not familiar with ansible as well as still provide a unified
+interface for installing the component (aligning with `oc cluster up` as it can
+leverage these OCP template files).
+
+Within the openshift-ansible role, installing in a way similar to APBs would
+create an interface that would not need to change once the role is created,
+with the exception being if new information need be passed into the container.
+
+If hosted components adhere to the structure required for the Service Broker
+to provision/deprovision bind/unbind it allows us to also open up the possibility
+to set up these components through the UI. This could further lighten the load
+placed on openshift-ansible and provide an alternative means to install hosted
+components after an OCP installation is completed.
+
+## User Story
+As a developer on OpenShift-Ansible,
+I want to roles for hosted components to run the APB container that is provided
+  by the developer of the component
+so that the majority of on-boarding and supporting a hosted component be the
+  responsibility of the developer of the component, reducing the amount of
+  copypasta, translating to Ansible, and changes within the openshift-ansible
+  role that is done as updates are made.
+
+## Acceptance Criteria
+* Verify that roles are simple and used to install|update|remove the component
+* Verify that roles leverage APB containers to maintain the objects
+  for the component
+* Verify that openshift-ansible roles are idempotent (side effects only when necessary)
+* Verify that the component APB containers adhere to necessary structure so that
+  it can be reused by the Serice Broker if desired
+* Verify that the APB role and playbook leverage oc and provided templates to
```

Comment:
I wouldn't mandate the template based install within the container, I think one benefit of the blackbox approach is that various teams familiar with different types of deployment approaches will be able to provide installers for their components that they are able to easily maintain.

Reply:
My thought behind 'mandating' it was so that `oc cluster up` can reuse them. Not sure if that's an accurate thought though. It'd be convenient if `oc cluster up` could use the APBs directly.
cc @csrwng

Reply:
With @jcantrill's work, cluster up now uses containerized ansible playbooks to deploy metrics and logging, and that's actually the preferred method of deploying things going forward. What the playbooks use to provision things is a blackbox, which is nice.

Reply:
Thanks! I'll update this then from a mandate. Maybe it can be mentioned as an approach for setting up the APB role to reduce replication of code for some components.